pkg/spoke/addon/registration_controller.go

```diff
@@ -0,0 +1,232 @@
+package addon
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/openshift/library-go/pkg/controller/factory"
+	"github.com/openshift/library-go/pkg/operator/events"
+	operatorhelpers "github.com/openshift/library-go/pkg/operator/v1helpers"
+	certificatesv1beta1 "k8s.io/api/certificates/v1beta1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/informers"
+	certificatesinformers "k8s.io/client-go/informers/certificates/v1beta1"
+	"k8s.io/client-go/kubernetes"
+	csrclient "k8s.io/client-go/kubernetes/typed/certificates/v1beta1"
+	"k8s.io/klog/v2"
+
+	addoninformerv1alpha1 "github.com/open-cluster-management/api/client/addon/informers/externalversions/addon/v1alpha1"
+	addonlisterv1alpha1 "github.com/open-cluster-management/api/client/addon/listers/addon/v1alpha1"
+	"github.com/open-cluster-management/registration/pkg/clientcert"
+)
+
+// addOnRegistrationController monitors ManagedClusterAddOn crs on hub and starts addOn registration
+// according to the registration configuration from cr annotations.
+type addOnRegistrationController struct {
+	clusterName    string
+	agentName      string
+	kubeconfigData []byte
+	kubeClient     kubernetes.Interface
+	hubAddOnLister addonlisterv1alpha1.ManagedClusterAddOnLister
+	hubCSRInformer certificatesinformers.CertificateSigningRequestInformer
+	hubCSRClient   csrclient.CertificateSigningRequestInterface
+	recorder       events.Recorder
+
+	// registrationConfigs maps the addon name to a slice of registrationConfigs
+	addOnRegistrationConfigs map[string][]registrationConfig
+	// stopRegistrationFuncs maps the key of a registrationConfig to its stopRegistrationFunc
+	stopRegistrationFuncs map[string]context.CancelFunc
+}
+
+// NewAddOnRegistrationController returns an instance of addOnRegistrationController
+func NewAddOnRegistrationController(
+	clusterName string,
+	agentName string,
+	kubeconfigData []byte,
+	kubeClient kubernetes.Interface,
+	hubCSRInformer certificatesinformers.CertificateSigningRequestInformer,
+	hubAddOnInformers addoninformerv1alpha1.ManagedClusterAddOnInformer,
+	hubCSRClient csrclient.CertificateSigningRequestInterface,
+	recorder events.Recorder,
+) factory.Controller {
+	c := &addOnRegistrationController{
+		clusterName:              clusterName,
+		agentName:                agentName,
+		kubeconfigData:           kubeconfigData,
+		kubeClient:               kubeClient,
+		hubAddOnLister:           hubAddOnInformers.Lister(),
+		hubCSRInformer:           hubCSRInformer,
+		hubCSRClient:             hubCSRClient,
+		recorder:                 recorder,
+		addOnRegistrationConfigs: map[string][]registrationConfig{},
```

Comment:
why is this a single value instead of a slice since addons can logically have multiple registrationConfigs where only one is active